airbyte-integrations/connectors/source-linnworks/source_linnworks/source.py

```diff
@@ -0,0 +1,324 @@
+
+        self._authenticator = authenticator
+        self.start_date = start_date
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        if response.status_code != requests.codes.ok:
+            return None
+
+        json = response.json()
+        if not isinstance(json, list):
+            json = [json]
+        for record in json:
+            yield normalize(record)
+
+
+class LinnworksGenericPagedResult(ABC):
+    @abstractmethod
+    def paged_result(self, response: requests.Response) -> Mapping[str, Any]:
+        pass
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        result = self.paged_result(response)
+        del result["Data"]
+        result = normalize(result)
+
+        if result["page_number"] < result["total_pages"]:
+            result["page_number"] += 1
+            return result
+
+
+class Location(LinnworksStream):
+    # https://apps.linnworks.net/Api/Method/Locations-GetLocation
+    # Response: StockLocation https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Locations-ClassBase-StockLocation
+    # Allows 150 calls per minute
+    primary_key = "stock_location_int_id"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "/api/Locations/GetLocation"
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        return stream_state
+
+
+class StockLocations(LinnworksStream):
+    # https://apps.linnworks.net/Api/Method/Inventory-GetStockLocations
+    # Response: List<StockLocation> https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockLocation
+    # Allows 150 calls per minute
+    primary_key = "stock_location_int_id"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "/api/Inventory/GetStockLocations"
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None
+    ) -> Iterable[Mapping[str, Any]]:
+        records = super().read_records(sync_mode, cursor_field, stream_slice, stream_state)
+
+        for record in records:
+            location = Location(authenticator=self.authenticator)
```

Comment:
It looks strange to me that you're calling another stream from this stream. Do you think its mandatory? It looks like early data enrichment / normalization that could be performed downstream with eventual Airbyte normalization.

Reply:
Endpoint [GetStockLocations](https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockLocation) returns an array of locations with limited properties of the `StockLocation` entity. While endpoint [GetLocation](https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Locations-ClassBase-StockLocation) return all the properties for particular `StockLocationIntId`.

My goal is to extract all properties of all locations. N+1 requests are the only way to do that. However, it might be implemented in other ways. I haven't found a relevant example for the `full_refresh` stream in other connectors, while there are some for `incremental` streams utilizing `stream_slices` method, which is unavailable for `full_refresh`.

Can you please briefly describe how my desired functionality can be properly implemented or point to a relevant example in other connectors?

Reply:
For the same matters I explained in [this comment](https://github.com/airbytehq/airbyte/pull/7588#discussion_r743694543) I feel like you're altering the original response from the `GetStockLocations` endpoint. Nevertheless according to Linnworks api documentation there's no endpoint to get a list of all the locations. If such an endpoint existed I would have suggested that this enrichment operation should be done I a downstream `join` transformation. If you think this is the only way of retrieving the `StockLocation` metadata I agree with your way of doing this but I suggest you document why you're doing this.

Reply:
And I would also suggest to store the `location` data in a nested field rather than at the top level of the `StockLocation` schema.